drivers/dma/dma_silabs_siwx917.c

```diff
@@ -90,11 +104,153 @@ static inline int siwx917_dma_addr_adjustment(uint32_t adjustment)
-			      uint32_t channel, struct dma_config *config,
-			      UDMA_Channel_Info *channel_info)
+/* Sets up the scatter-gather descriptor table for a DMA transfer */
+static int siwx917_sg_fill_desc(RSI_UDMA_DESC_T *descs, const struct dma_config *config)
+{
+	const struct dma_block_config *block_addr = config->head_block;
+	RSI_UDMA_CHA_CONFIG_DATA_T *cfg;
+
+	for (int i = 0; i < config->block_count; i++) {
+		sys_write32((uint32_t)&descs[i].vsUDMAChaConfigData1, (mem_addr_t)&cfg);
+		/* Set the source and destination end addresses */
+		descs[i].pSrcEndAddr =
+			(void *)(block_addr->source_address +
+				 (block_addr->block_size - config->source_data_size));
+		descs[i].pDstEndAddr = (void *)(block_addr->dest_address +
+						(block_addr->block_size - config->dest_data_size));
+
+		/* Set the source and destination data sizes */
+		cfg->srcSize = siwx917_data_width(config->source_data_size);
+		cfg->dstSize = siwx917_data_width(config->dest_data_size);
+
+		/* Calculate the number of DMA transfers required */
+		if (block_addr->block_size / config->source_data_size > DMA_MAX_TRANSFER_COUNT) {
+			return -EINVAL;
+		}
+
+		/* Set the total number of DMA transfers */
+		cfg->totalNumOfDMATrans = block_addr->block_size / config->source_data_size - 1;
+
+		/* Set the transfer type based on whether it is a peripheral request */
+		if (siwx917_transfer_direction(config->channel_direction) == 1) {
+			cfg->transferType = UDMA_MODE_PER_ALT_SCATTER_GATHER;
+		} else if (siwx917_transfer_direction(config->channel_direction) == 0) {
+			cfg->transferType = UDMA_MODE_MEM_ALT_SCATTER_GATHER;
+		}
+
+		/* Set the arbitration size */
+		cfg->rPower = ARBSIZE_1;
+
+		if (siwx917_addr_adjustment(block_addr->source_addr_adj) < 0 ||
+		    siwx917_addr_adjustment(block_addr->dest_addr_adj) < 0) {
+			return -EINVAL;
+		}
+
+		/* Set source and destination address increments */
+		if (siwx917_addr_adjustment(block_addr->source_addr_adj) == UDMA_ADDR_INC_NONE) {
+			cfg->srcInc = UDMA_SRC_INC_NONE;
+		} else if (siwx917_addr_adjustment(block_addr->source_addr_adj) == 0) {
+			cfg->srcInc = siwx917_data_width(config->source_data_size);
+		}
+
+		if (siwx917_addr_adjustment(block_addr->dest_addr_adj) == UDMA_ADDR_INC_NONE) {
+			cfg->dstInc = UDMA_DST_INC_NONE;
+		} else if (siwx917_addr_adjustment(block_addr->dest_addr_adj) == 0) {
+			cfg->dstInc = siwx917_data_width(config->dest_data_size);
+		}
+
+		/* Move to the next block */
+		block_addr = block_addr->next_block;
+	}
+
+	if (block_addr != NULL) {
+		/* next_block address for last block must be null */
+		return -EINVAL;
+	}
+
+	/* Set the transfer type for the last descriptor */
+	if (siwx917_transfer_direction(config->channel_direction) == 1) {
+		descs[config->block_count - 1].vsUDMAChaConfigData1.transferType = UDMA_MODE_BASIC;
+	} else if (siwx917_transfer_direction(config->channel_direction) == 0) {
+		descs[config->block_count - 1].vsUDMAChaConfigData1.transferType = UDMA_MODE_AUTO;
+	}
```

Comment:
This is what I mean by hard to read code. Why do you want these `1` and `0` magic numbers in your code? Just use proper defines or an enum. Furthermore, you're calling the function twice which seems unnecessary, and also (again) missing the handling of the `else` case. Seems like this might work better as a switch statement:
```c
switch (siwx917_transfer_direction(config->channel_direction)) {
case TRANSFER_TO_OR_FROM_PER:
	descs[config->block_count - 1].vsUDMAChaConfigData1.transferType = UDMA_MODE_BASIC;
	break;
case TRANSFER_MEM_TO_MEM:
	descs[config->block_count - 1].vsUDMAChaConfigData1.transferType = UDMA_MODE_AUTO;
	break;
default:
	return -EINVAL;
}
```